src/Workspaces/Core/Portable/Rename/Renamer.SyncNamespaceDocumentAction.cs

```diff
@@ -0,0 +1,121 @@
+            }
+
+            public override string GetDescription(CultureInfo? culture)
+             => WorkspacesResources.ResourceManager.GetString("Sync_namespace_to_folder_structure", culture ?? WorkspacesResources.Culture)!;
+
+            internal override async Task<Solution> GetModifiedSolutionAsync(Solution solution, CancellationToken cancellationToken)
+            {
+                solution = solution.WithDocumentFolders(DocumentId, _newFolders);
+
+                if (!_analysis.SupportsSyncNamespace)
+                {
+                    return solution;
+                }
+
+                var document = solution.GetRequiredDocument(DocumentId);
+                var changeNamespaceService = document.GetRequiredLanguageService<IChangeNamespaceService>();
+
+                var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+                var targetNamespace = BuildNamespaceFromFolders(document.Folders, syntaxFacts);
+
+                var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                // Don't descend into anything other than top level declarations from the root.
+                // ChangeNamespaceService only controls top level declarations right now
+                var namespaceDeclarations = syntaxRoot.DescendantNodes(n => !syntaxFacts.IsDeclaration(n)).Where(n => syntaxFacts.IsNamespaceDeclaration(n));
+
+                foreach (var namespaceDeclaration in namespaceDeclarations)
+                {
+                    solution = await changeNamespaceService.ChangeNamespaceAsync(
+                        document,
+                        namespaceDeclaration,
+                        targetNamespace,
+                        cancellationToken).ConfigureAwait(false);
+
+                    document = solution.GetRequiredDocument(document.Id);
+                }
+
+                return solution;
+
+                static string BuildNamespaceFromFolders(
+                    IEnumerable<string> folders,
+                    ISyntaxFacts syntaxFacts)
+                {
+                    var parts = folders.SelectMany(folder => folder.Split(new[] { '.' }));
+
+                    // string.Empty is used to represent the global namespace
+                    // when calling ChangeNamespaceAsync
+                    return parts.All(syntaxFacts.IsValidIdentifier) ? string.Join(".", parts) : string.Empty;
+                }
+            }
+
+            public static Task<SyncNamespaceDocumentAction> CreateAsync(Document document, IReadOnlyList<string> newFolders, OptionSet optionSet, CancellationToken _)
+            {
+                var errors = new ArrayBuilder<string>();
+
+                var analysisResult = Analyze(document);
+
+                if (!analysisResult.SupportsSyncNamespace)
+                {
+                    errors.Add(WorkspacesResources.The_project_does_not_support_sync_namespace);
```

Comment:
So what is supposed to deal with this? Imagining the project system side of this: if I move a file around, I don't want to see an error that says "we tried to apply this but decided we didn't support it". I would just imagine CreateAsync() never made a SyncNamespaceDocumentAction at all in that case. We don't want the project system having to somehow know if this is safe to call or not.

Reply:
Changed to not returning an action